src/lsm/scan_tree.zig

```diff
@@ -0,0 +1,666 @@
+fn ScanTreeLevelType(comptime ScanTree: type, comptime Storage: type) type {
+    return struct {
+        const ScanTreeLevel = @This();
+        const LevelTableValueBlockIterator = LevelTableValueBlockIteratorType(
+            ScanTree.Table,
+            Storage,
+        );
+        const BlockPtr = ScanTree.Grid.BlockPtr;
+        const BlockPtrConst = ScanTree.Grid.BlockPtrConst;
+
+        const TableInfo = ScanTree.TableInfo;
+        const Manifest = ScanTree.Manifest;
+
+        const LevelBuffer = ScanTree.ScanBuffer.LevelBuffer;
+
+        const Table = ScanTree.Table;
+        const Key = Table.Key;
+        const Value = Table.Value;
+        const key_from_value = Table.key_from_value;
+
+        scan: *ScanTree,
+        level_index: u8,
+        iterator: LevelTableValueBlockIterator,
+        buffer: LevelBuffer,
+
+        /// State over the manifest.
+        manifest: union(enum) {
+            begin,
+            next_table: ScanTree.Key,
+            finished,
+        },
+
+        /// State over the values.
+        values: union(enum) {
+            fetching,
+            buffered: RangeCursor,
+            finished,
+        },
+
+        tracer_slot: ?tracer.SpanStart = null,
+
+        pub inline fn init(scan: *ScanTree, buffer: LevelBuffer, level_index: u8) ScanTreeLevel {
+            return .{
+                .level_index = level_index,
+                .scan = scan,
+                .iterator = LevelTableValueBlockIterator.init(),
+                .buffer = buffer,
+                .manifest = .begin,
+                .values = .fetching,
+            };
+        }
+
+        /// Moves the level iterator to the next `table_info` that might contain the key range.
+        pub fn next_table(level: *ScanTreeLevel) void {
+            const scan: *ScanTree = level.scan;
+
+            const table_info: ?*const TableInfo = blk: {
+                const manifest: *Manifest = &scan.tree.manifest;
+                var key_exclusive: ?Key = switch (level.manifest) {
+                    .begin => null,
+                    .next_table => |key| key,
+                    .finished => unreachable,
+                };
+
+                if (manifest.next_table(.{
+                    .level = level.level_index,
+                    .snapshot = scan.snapshot,
+                    .key_min = scan.key_min,
+                    .key_max = scan.key_max,
+                    .key_exclusive = key_exclusive,
+                    .direction = scan.direction,
+                })) |next_table_info| {
+                    key_exclusive = switch (scan.direction) {
+                        .ascending => next_table_info.key_max,
+                        .descending => next_table_info.key_min,
+                    };
+
+                    if (key_exclusive.? < scan.key_min or
+                        key_exclusive.? > scan.key_max)
+                    {
+                        level.manifest = .finished;
```

Comment:
Yes, because `next_table` was called with `key_exclusive_before`, and now we need to check if the next table contains `key_exclusive_after`, otherwise we finished.

It wasn't obvious when we used the same variable `key_exclusive` for both, now it's self-explanatory :+1: